quarterback-packages/api/src/routes/doc/doc.ctrl.ts

```diff
@@ -101,3 +103,150 @@ export const deleteDocument = async (
     next(err)
   }
 }
+
+export type DocData = {
+  steps: Step[]
+  version: number
+  clientIDs: number[]
+  doc: any
+}
+const tempDocs = new Map<string, DocData>()
+const getGlobalTemp = () => {
+  return tempDocs
+}
+
+const getDocHistory = (docId: string) => {
+  const temp = getGlobalTemp()
+  const docData = temp.get(docId)
+  if (!docData) {
+    const template = {
+      steps: [], // steps have to be dropped at some point
+      clientIDs: [], // needs to be dropped along with steps
+      version: 0,
+      doc: undefined,
+    }
+    temp.set(docId, template)
+    return template
+  }
+  return docData
+}
+
+export const receiveSteps = async (
+  req: AuthRequest<any>,
+  res: AuthResponse,
+  next: NextFunction
+) => {
+  try {
+    const { documentId } = req.params
+    const newSteps = req.body.steps
+    const clientID = req.body.clientID // clientID is not the same as userID, it's an ID of the client exactly (app instance running in the browser)
+    const clientVersion = req.body.version // client the version on top of which it updates, in other words it says last version known to it
+
+    const docSearch = await docService.findDocument(documentId)
+    if ('data' in docSearch) {
+      // @ts-ignore
+      const { version } = docSearch.data
+
+      // this is needed, commented for testing only
+      if (version != clientVersion) {
+        /*
+        if client is based on an older version we deny his request. At some point in time client should receive update and will then
+        resend his steps after rebasing them. Client will persist its steps.
+        */
+        return
+      }
+
+      const cachedItem = getDocHistory(documentId)
+
+      let pmDoc = cachedItem.doc || schema.nodeFromJSON(docSearch.data.doc)
+
+      newSteps.forEach((jsonStep) => {
+        const step = Step.fromJSON(schema, jsonStep)
+        pmDoc = step.apply(pmDoc).doc || pmDoc
+        cachedItem.steps.push(step)
+        cachedItem.clientIDs.push(clientID)
+      })
+      cachedItem.version = clientVersion
+      cachedItem.doc = pmDoc
+
+      // docService.updateDocument(documentId, {doc: pmDoc.toJSON()})
+      res.sendStatus(200)
+
+      // Signal server side listener
+      sendStepsToAll({ steps: newSteps, clientIDs: cachedItem.clientIDs, version: clientVersion })
+    } else {
+      next(new CustomError('Unable to save steps', 500))
+    }
+  } catch (err) {
+    next(err)
+  }
+}
+
+function sendStepsToAll(data: unknown) {
+  global.clients.forEach((client) => client.res.write(`data: ${JSON.stringify(data)}\n\n`))
```

Comment:
@mbetamony this is dispatching function for currently open SSE connections.

Reply:
We should have a check here to determine if the client is part of a specific document, right? So, if we have two documents with the following IDs: document1 and document2, and we're working with document1, we'll be sending data to doucment1 and document2 clients which is wrong, correct?

Reply:
Indeed we need to check permissions based on client authorisation/role for a specific document. I would say we need to check it but not (maybe not only) in the dispatch but primarily while client establishes the SSE connection. Maybe here it's also a good idea to check it in case the user's role has changed, but if it would involve querying DB to check the role for each listener - it may impede performance, so in my opinion checking on connection open should be enough for now.